src/qs8-gemm/c8-wasmdot.c.in

```diff
@@ -0,0 +1,342 @@
+        c${M} = c${M-1};
+      }
+    $else:
+      if XNN_UNPREDICTABLE(mr < ${M+1}) {
+        a${M} = a${M-1};
+        c${M} = c${M-1};
+      }
+
+  $if DATATYPE == "QD8":
+    const v128_t vmin = wasm_v128_load32_splat(&params->scalar.min);
+    const v128_t vmax = wasm_v128_load32_splat(&params->scalar.max);
+    XNN_FORCE_REALIZATION(vmin);
+    XNN_FORCE_REALIZATION(vmax);
+  $else:
+    $if DATATYPE != "QC8":
+      const v128_t vscale = wasm_v128_load32_splat(&params->fp32_scalar.scale);
+      XNN_FORCE_REALIZATION(vscale);
+
+    const v128_t vmagic_bias = wasm_f32x4_const_splat(12582912.0f);
+    const int32_t output_min_less_zero_point = (int32_t) params->fp32_scalar.output_min - (int32_t) params->fp32_scalar.output_zero_point;
+    const v128_t vmagic_min = wasm_i32x4_splat((int32_t) float_as_uint32(12582912.0f + output_min_less_zero_point));
+    const v128_t vmagic_bias_less_output_zero_point = wasm_i32x4_splat(INT32_C(0x4B400000) - (int32_t) params->fp32_scalar.output_zero_point);
+    const v128_t voutput_max = wasm_i8x16_splat(params->fp32_scalar.output_max);
+    XNN_FORCE_REALIZATION(vmagic_bias);
+    XNN_FORCE_REALIZATION(vmagic_min);
+    XNN_FORCE_REALIZATION(vmagic_bias_less_output_zero_point);
+    XNN_FORCE_REALIZATION(voutput_max);
+
+  $if not SDOT:
+    const v128_t vsign_mask = wasm_u8x16_const_splat(UINT8_C(0x80));
+  do {
+    $if DATATYPE == "QD8":
+      $for M in range(MR):
+        $if SDOT:
+          const v128_t vinput_zero_point${M} = wasm_v128_load32_splat(&quantization_params[${M}].zero_point);
+        $else:
+          const v128_t vinput_zero_point${M} = wasm_i32x4_splat((int32_t) quantization_params[${M}].zero_point + 128);
+      $for N in range(0, NR, 4):
+        v128_t vksum${ABC[N:N+4]} = wasm_v128_load(w); w = (const int32_t*) w + 4;
+        $for M in range(MR):
+          v128_t vsum${M}x${ABC[N:N+4]} = wasm_i32x4_mul(vksum${ABC[N:N+4]}, vinput_zero_point${M});
+          v128_t vacc${M}x${ABC[N:N+2]} = wasm_u64x2_extend_low_u32x4(vsum${M}x${ABC[N:N+4]});
+          v128_t vacc${M}x${ABC[N+2:N+4]} = wasm_u64x2_extend_high_u32x4(vsum${M}x${ABC[N:N+4]});
+    $else:
+      $for N in range(0, NR, 2):
+        v128_t vacc0x${ABC[N:N+2]} = wasm_u64x2_load32x2(w); w = (const int32_t*) w + 2;
+      $for N in range(0, NR, 2):
+        $for M in range(1, MR):
+          v128_t vacc${M}x${ABC[N:N+2]} = vacc0x${ABC[N:N+2]};
+
+    size_t k = kc;
+    while (${"k >= 16 * sizeof(int8_t)" if UNROLL else "k != 0"}) {
+      $for M in range(MR):
+        $if SDOT:
+          const v128_t va${M}x01 = wasm_v128_load64_splat(a${M});
+          $if UNROLL:
+            const v128_t va${M}x23 = wasm_v128_load64_splat((const int8_t*) a${M} + 8);
+        $else:
+          const v128_t va${M}x01 = wasm_v128_xor(wasm_v128_load64_splat(a${M}), vsign_mask);
+          $if UNROLL:
+            const v128_t va${M}x23 = wasm_v128_xor(wasm_v128_load64_splat((const int8_t*) a${M} + 8), vsign_mask);
```

Comment:
on native I do it this way to allow embedded broadcast of 64 bits for xor.  But if you dont have embedded broadcast, it may be worth trying the opposite order
load a full vector
xor a full vector
use 2 shuffles to isolate 64 bits
For float kernels, we call this 'dup'.  e.g. load 4 floats and use dup to isolate 1 float.  Its usually the worst of the methods.  A small improvement is C8S2 where multiply the full vector of 16, and then after the multiply, shuffle... which would swap the upper 8 bytes with lower 8 bytes.
So instead of shuffle+mul+shuffle+mul its mul+shuffle+mul